### DISCUSSION TITLE
Help Importing Own Labels

### DISCUSSION THREAD

**Discussion by dpardi02** [Project-MONAI/MONAILabel#1540] [Q&A]
Hello. I am using Monai Label to do segmentation of scans that I have generated. I am attempting to train the set on labels that I have already created manually and that I know are correct. However, I am running into some issues in doing so. Although I select the label in the "Import Label" section from the tools drop-down, when I go to submit it to the server it does not appear in the workspace. For clarification, I am using Monai Label within 3dslicer. I am not very familiar with the segmentation and deep edit process so I am sure there are aspects to it that I am missing. My question stems from the fact that when I upload the label as the main data entry and then apply it to itself, it appears. A segmentation region shows up like one would if I had made it manually within Monai Label. However, when I do this same process with the original micro CT scan in nifti format, the submit process does not work and no label appears. Any help is greatly appreciated and I am happy to answer any questions that would aid in the understanding of my issue. Thank you so much.

**Comment by dpardi02** [comment]
My follow-up question to this is whether the imported labels have to be the label with the scan or if I can upload the label and the scan as two separate files.

**Comment by tangy5** [comment]
There are several thing you could double check:

1. If you will directly use the existing labels for training and just use monai label to trigger the training process, you could manually move the label files in your studies folder: /<studies_folder>/labels/final, the "final" folder is where monailabel save the final annotations and use for training. 
2. If you would like to use the your original label in 3D Slicer and further edits then save the label, you could save the label files in  /<studies_folder>/labels/original folder, this is the place where monailabel can load the beginning label and show in 3D Slicer.
3. The imported label, you will need to define yourself in the model scripts. For example: like the "segmentation" sample model in the radiology sample app, you could define your own labels here: https://github.com/Project-MONAI/MONAILabel/blob/c90f42c0730554e3a05af93645ae84ccdcb5e14b/sample-apps/radiology/lib/configs/segmentation.py#L37

**Reply by dpardi02** [reply]
Thank you for the idea. I will give it a try. Are there any nomenclature/file type specifications I should be following when putting the labels in that folder? For example, my current labels define one region and then everything else as "other" and are just binary, black and white volumes saved as niftis. Would this suffice?

**Reply by diazandr3s** [reply]
You could use any segment name. The most important thing here is to consistently use the same indexes or numbers representing the segments. For instance, if you want to segment and spleen (liver=1, spleen=2)
You have to make sure the liver is always 1 and the spleen is always 2 in the annotations.

Hope this helps,

**Reply by dpardi02** [reply]
Got it. Thank you. I think I have the general idea but am still encountering an error. For testing purposes, I have been using a dataset with one volume and one label file. However, when I try the suggestions above I have the following problems.

Label in the "Final" label folder: When I put the label here as a gzip file, Monai Label shows the active learning at 100% which makes sense but then when I go to train it always stays at 0% for accuracy. The labels are not perfect as we are still working through this process but I surely expected some type of accuracy for this step.

Label in the "Original" label folder: When the label is entered here in the same format and "Next Sample" is clicked within Monai Label, the volume will load however, there is no existing label shown. 

I am assuming there is an issue with my label submission in this part of the process. I have changed the segmentation config file to only include two self-labels and have turned the pre-trained model off, but is there any other reason my label is not being accepted? Thank you for your continued help.

**Reply by diazandr3s** [reply]
The 0% accuracy is because there are no "validation" samples to check for the model accuracy. I understand you only have one training sample at the moment. By default, 20% of the training samples are separated for validation/test.

**Reply by dpardi02** [reply]
I see. So in theory it should work as I add more samples to my set. Thank you so much for your help.

**Comment by longchingkwok331** [comment]
Hi, I have used the 3D slicer's lung CT segmenter module to segment lung regions and submitted to the MONAI label server.  However, the labels are not stored in the dataset directory. I use the command _$ monailabel start_server --app apps/radiology --studies \data-for-testing --conf models segmentation_ to start the server and I have edited my segmentation file accordingly as shown in the photo. May I know where should the labels be stored ? 
![image](https://github.com/Project-MONAI/MONAILabel/assets/138690628/181ef605-736b-485b-8812-b4fcce5f187b)

May I know if 3D slicer created a lung segmentation that is structured as in the picture, should I put the labels in the segmentation same as that structure? e.g. "right lung" : 1, 
"left lung": 2, "thoracic cavity":  3 ......
The "segmentation_3000522: Unnamed Series" column in the picture is produced after I submit the labels and I am not sure why 
![image](https://github.com/Project-MONAI/MONAILabel/assets/138690628/6f48161f-caa9-4687-b83d-2810c81af291)

thank you

**Reply by diazandr3s** [reply]
Hi @longchingkwok331,

Thanks for asking this.

You could also upload the existing annotation as it is shown here: https://youtu.be/KjwuFx0pTXU?si=8wHgF6SB3FuxjbbL&t=173
Make sure the number of segments and indices are consistent with the started radiology app.

Otherwise, I'd suggest you first annotate volumes for training, and make sure the indices are consistent in all the volumes (i.e. left lung is always 2, right lung is always 1, etc) and then start the MONAI Label server pointing to those annotated volumes.

Hope this helps,

**Reply by longchingkwok331** [reply]
thank you. "Make sure the number of segments and indices are consistent with the started radiology app." How should we check the indices of the segments in 3D slicer? 
For example, I have right lung, left lung, thoracic cavity, lungs..... in my attached screenshot, then the indices will be right lung:1, left lung:2 , thoracic cavity:3 , lungs:4? 

thank you

**Reply by diazandr3s** [reply]
Hi @longchingkwok331,

Indices have to be checked before training. Sometimes they are not consecutive and that must be reflected in the configs file. 
It depends on how you created the segmentations or where you get them.

A simple MONAI script can show you the indices on each NIfTI file:
```

import glob
import os
import numpy as np
from monai.transforms import LoadImage

source_labels = "/PATH_TO_SEGMENTATION_FILES/"

all_files = glob.glob(os.path.join(source_labels, "*.nii.gz"))

for idx, img_path in enumerate(all_files):
    d = LoadImage(image_only=True)(img_path)
    dirname, file = os.path.split(img_path)
    # print(f"Spacing: {np.diag(d.meta.get('affine'))} and name: {file}")
    print(f"Image: {file} => max intensity {d.array.max()} - min intensity {d.array.min()}")
    print("Unique values: ", np.unique(d.array))
    print(f"Labelmap size: {d.array.shape}")
```
I hope this helps
